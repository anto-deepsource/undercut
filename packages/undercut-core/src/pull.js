export * from "./pull_core";
export * from "./pull_operations";
export * from "./pull_targets";
