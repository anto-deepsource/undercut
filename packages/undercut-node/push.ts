// Temporary file while there're still issues in Jest/TS/etc with package exports.
export * from "@undercut/push";
