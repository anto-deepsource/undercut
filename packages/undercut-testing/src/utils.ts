import type { Observer } from "@undercut/types";

import { asObserverFactory } from "@undercut/utils";

export function getArrayObserver<T>(array: Array<T>) {
	return asObserverFactory(function* (): Observer<T> {
		while (true) {
			array.push(yield);
		}
	})();
}
