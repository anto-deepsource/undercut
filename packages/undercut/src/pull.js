export * from "./pull/pull_core.js";
export * from "./pull/pull_operations.js";
export * from "./pull/pull_targets.js";
