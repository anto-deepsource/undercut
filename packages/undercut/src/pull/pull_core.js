import { assert, assertSource } from "../utils/assertions.js";
import { isFunction, isIterable } from "../utils/lang.js";

import Iterable from "../iterable.js";

const operationErrorMessage = "An operation must be a function taking an iterable as an argument and returning another iterable.";

function operationsToIterator(operations, source) {
	let result = source;

	for (const operation of operations) {
		assert(isFunction(operation), operationErrorMessage);

		result = operation(result);

		assert(isIterable(result), operationErrorMessage);
	}

	if (result === source) {
		return source[Symbol.iterator]();
	}

	return result;
}

export function pullLine(operations, source) {
	assert(isIterable(operations), "Operations are required, could be an array or another iterable.");
	assertSource(source);

	return new Iterable(() => operationsToIterator(operations, source));
}

export function pull(target, operations, source) {
	assert(isFunction(target), "Target is required, must be a function accepting an iterable.");
	assert(isIterable(operations), "Operations are required, could be an array or another iterable.");
	assertSource(source);

	return target(operationsToIterator(operations, source));
}
