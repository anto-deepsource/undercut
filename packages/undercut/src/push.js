export * from "./push/push_core.js";
export * from "./push/push_operation.js";
// export * from "./push/push_sources.js";
export * from "./push/push_target.js";
