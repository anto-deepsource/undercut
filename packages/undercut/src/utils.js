import * as compare from "./utils/compare.js";

export { compare };
export * from "./utils/function.js";
export * from "./utils/iterable.js";
export * from "./utils/language.js";
