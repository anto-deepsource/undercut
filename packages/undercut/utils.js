import * as compare from "./src/utils/compare.js";

export { compare };
export * from "./src/utils/function.js";
export * from "./src/utils/helpers.js";
export * from "./src/utils/lang.js";
