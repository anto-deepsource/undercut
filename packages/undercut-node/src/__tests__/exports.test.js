import { describe, expect, test } from "@jest/globals";

import * as pullExports from "@undercut/pull";
import * as pushExports from "@undercut/push";
import * as utilsExports from "@undercut/utils";

import * as myPullExports from "../../exports/pull.js";
import * as myPushExports from "../../exports/push.js";
import * as myUtilsExports from "../../exports/utils.js";

import * as myOwnExports from "../index.js";

describe(`Exports of @undercut/node`, () => {
	test(`should be stable`, () => {
		expect(Object.keys(myOwnExports).sort()).toMatchSnapshot();
	});

	test(`/pull should match @undercut/pull`, () => {
		expect(Object.keys(myPullExports).sort()).toEqual(Object.keys(pullExports).sort());
	});

	test(`/push should match @undercut/push`, () => {
		expect(Object.keys(myPushExports).sort()).toEqual(Object.keys(pushExports).sort());
	});

	test(`/utils should match @undercut/utils`, () => {
		expect(Object.keys(myUtilsExports).sort()).toEqual(Object.keys(utilsExports).sort());
	});
});
