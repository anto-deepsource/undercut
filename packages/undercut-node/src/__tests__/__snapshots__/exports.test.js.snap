// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exports of @undercut/node should be stable 1`] = `Array []`;
